File: Array/Parenthesis/Remove_Invalid_Parentheses.ts
// Leetcode: https://leetcode.com/problems/remove-invalid-parentheses/

const isValid = (s: string): boolean => {
    const stack: string[] = [];
    for (const c of s) {
        if (c === "(") {
            stack.push("(");
        } else if (c === ")") {
            if (stack.length > 0) {
                stack.pop();
            } else {
                return false;
            }
        }
    }
    return stack.length === 0;
};

const removeInvalidParentheses = (s: string): string[] => {
    let result: string[] = [];
    let visit: Set<string> = new Set();
    let queue = [s];
    let found = false;

    while (queue.length) {
        let node = queue.shift() as string;
        if (isValid(node)) {
            result.push(node);
            found = true;
        }
        if (!found) {
            for (let i = 0; i < node.length; i++) {
                if ("()".includes(node[i])) {
                    let new_node = node.slice(0, i) + node.slice(i + 1);
                    if (!visit.has(new_node)) {
                        queue.push(new_node);
                        visit.add(new_node);
                    }
                }
            }
        }
    }

    return result;
};

File: Array/Parenthesis/Remove_Invalid_Parentheses.py
#  Leetcode: https://leetcode.com/problems/remove-invalid-parentheses/

# Solution 1: BFS
class Solution:
    def is_valid(s: str) -> bool:
        stack = []
        for c in s:
            if c == "(":
                stack.append("(")
            elif c == ")":
                if len(stack) > 0:
                    stack.pop()
                else:
                    return False
        return len(stack) == 0

    def removeInvalidParentheses(self, s: str) -> List[str]:
        result = []
        visit = set()
        queue = [s]
        found = False

        while queue:
            node = queue.pop(0)
            if Solution.is_valid(node):
                result.append(node)
                found = True
            if not found:
                for i in range(len(node)):
                    if node[i] in "()":
                        new_node = node[:i] + node[i+1:]
                        if new_node not in visit:
                            queue.append(new_node)
                            visit.add(new_node)

        return result

# Solution 2: Recursion
class Solution:
    def get_invalid_paranthesis_count(s: str) -> int:
        stack = []
        for c in s:
            if c=='(':
                stack.append(c)
            elif c==')':
                if stack and stack[-1]=='(':
                    stack.pop()
                else:
                    stack.append(c)
        return len(stack)

    def removeInvalidParentheses(self, s: str) -> List[str]:
        visit = set()    
        count = Solution.get_invalid_paranthesis_count(s)
        
        @cache
        def recursion(st: str, count: int) -> None:
            if count == 0 and Solution.get_invalid_paranthesis_count(st) == 0:
                visit.add(st)
            else:
                for i in range(len(st)):
                    new_st = st[0:i] + st[i+1:]
                    recursion(new_st, count - 1)

        recursion(s, count)
        return visit